### PR TITLE
[DNM] Error middlewares

### PR DESCRIPTION
**Pull request by oleurud**
With the same idea of middlewarify PR, the objective of this PR is to break down the error middleware making it more modular using several middlewares and following the single responsibility principle.

With this changes, we will be able to integrate easily the new incoming features.

### CODE REVIEW

File: lib/cartodb/middleware/error/logger.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
+const debug = require('debug')('windshaft:cartodb:error-middleware:status-code');
```
[the remark this answers is not in the document]
Comment (oleurud) [reply lib/cartodb/middleware/error/logger.js]: Oh my god!

File: lib/cartodb/middleware/error/prepare.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+
+module.exports = function prepare () {
+    return function prepareMiddleware(err, req, res, next) {
+        let errors = Array.isArray(err) ? err : [err];
+
+        const label = err.label || 'UNKNOWN';
+        err = errors[0] || new Error(label);
```
[the remark this answers is not in the document]
Comment (oleurud) [reply lib/cartodb/middleware/error/prepare.js]: Yes, this part is in my head, we need to change. I like your idea

File: lib/cartodb/middleware/error/status-code.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+function statusFromErrorMessage(errMsg) {
+    // Find an appropriate statusCode based on message
+    // jshint maxcomplexity:7
+    let statusCode = 400;
+    if ( -1 !== errMsg.indexOf('permission denied') ) {
+        statusCode = 403;
+    }
+    else if ( -1 !== errMsg.indexOf('authentication failed') ) {
```
[the remark this answers is not in the document]
Comment (oleurud) [reply lib/cartodb/middleware/error/status-code.js]: Sure!

File: lib/cartodb/server.js [hunk 4]
```diff
@@ -46,7 +46,7 @@ var MapConfigAdapter = require('./models/mapconfig/adapter');
 var StatsBackend = require('./backends/stats');
 
 const lzmaMiddleware = require('./middleware/lzma');
-const errorMiddleware = require('./middleware/error-middleware');
```
[the remark this answers is not in the document]
Comment (oleurud) [reply lib/cartodb/server.js]: Thanks for helping me :)

**Comment by dgaubert** [comment]
By the way, I'd like to leave an example of what It's the approach that I'd like to implement step by step:

```js
const renderer = require('carto-raster-renderer');
const { TooManyRequests, ExpiredAuthApiTokenError } = renderer.errors;
const createError = require('http-errors');
const { NotFound, Unauthorized } = createError;

// tile.js
module.exports = function tile () {
    return function tileMiddleware (req, res, next) {
        const { id, z, x, y } = req.params;

        renderer.getMap(id).getTile(z, x, y, (err, tile) => {
            if (err) {
                return next(err);
            }

            if (tile === undefined) {
                return next(new NotFound());
            }

            res.status(200).send(tile);
        });
    }
}

// tile-error.js
module.exports = function tileError () {
    return function tileErrorMiddleware (err, req, res, next) {
        if (err instanceof RenderLimitError) {
            const error = new TooManyRequests('You are over platform limits');

            error.label = err.label;
            error.type = err.type;
            error.subtype = err.subtype;

            return next(error);
        }

        if (err instanceof ExpiredAuthApiTokenError) {
            const error = new Unauthorized('Your token has expired');

            error.label = err.label;
            error.type = err.type;
            error.subtype = err.type;

            return next(error);
        }

        next(err);
    };
};

// error.js
module.exports = function error () {
    return function errorMiddleware (err, req, res, next) {
        const errors = Array.isArray(err) ? err : [ err ];

        res.status(err.status).json({
            errors: errors.map(() => { /* get messages */ }),
            errors_with_context: errors.map(() => { /* get messages && context */ })
        });
    };
};
```

**Comment by oleurud** [comment]
I like very much your approach. It looks very well, it is easy to understand and maintain. 

I want to add a small point to the same solution. I think that probably, we will need more than 1 generic middleware because we can need to do the same in several middlewares (timeout limits for example)

**Comment by dgaubert** [comment]
@oleurud this PR is staled, do you think is worth to keep it open? Shall we close it?

**Comment by oleurud** [comment]
Yes, I would forget it. I like the main idea of the refactor, but It is too old and it still needs changes from your comments and our expertise from other related PRs. So, it would be easy to start from scratch than taking this one

**Pull request closed without merging**